flytekit/image_spec/default_builder.py

```diff
@@ -0,0 +1,245 @@
+    env_dict = {"PYTHONPATH": "/root", _F_IMG_ID: image_spec.image_name()}
+
+    if image_spec.env:
+        env_dict.update(image_spec.env)
+
+    env = " ".join(f"{k}={v}" for k, v in env_dict.items())
+
+    apt_packages = ["ca-certificates", "bzip2", "curl"]
+    if image_spec.apt_packages:
+        apt_packages.extend(image_spec.apt_packages)
+
+    apt_install_command = APT_INSTALL_COMMAND_TEMPLATE.substitute(APT_PACKAGES=" ".join(apt_packages))
+
+    if image_spec.source_root:
+        source_path = tmp_dir / "src"
+
+        ignore = IgnoreGroup(image_spec.source_root, [GitIgnore, DockerIgnore, StandardIgnore])
+        shutil.copytree(
+            image_spec.source_root,
+            source_path,
+            ignore=shutil.ignore_patterns(*ignore.list_ignored()),
+            dirs_exist_ok=True,
+        )
+        copy_command_runtime = "COPY --chown=flytekit ./src /root"
+    else:
+        copy_command_runtime = ""
+
+    conda_packages = image_spec.conda_packages or []
+    conda_channels = image_spec.conda_channels or []
+
+    if conda_packages:
+        conda_packages_concat = " ".join(conda_packages)
+    else:
+        conda_packages_concat = ""
+
+    if conda_channels:
+        conda_channels_concat = " ".join(f"-c {channel}" for channel in conda_channels)
+    else:
+        conda_channels_concat = ""
+
+    if image_spec.python_version:
+        python_version = image_spec.python_version
+    else:
+        python_version = f"{sys.version_info.major}.{sys.version_info.minor}"
+
+    if image_spec.commands:
+        run_commands = "\n".join(f"RUN {command}" for command in image_spec.commands)
+    else:
+        run_commands = ""
+
+    docker_content = DOCKER_FILE_TEMPLATE.substitute(
+        PYTHON_VERSION=python_version,
+        PYTHON_INSTALL_COMMAND=python_install_command,
+        CONDA_PACKAGES=conda_packages_concat,
+        CONDA_CHANNELS=conda_channels_concat,
+        APT_INSTALL_COMMAND=apt_install_command,
+        BASE_IMAGE=base_image,
+        ENV=env,
+        COPY_COMMAND_RUNTIME=copy_command_runtime,
+        RUN_COMMANDS=run_commands,
+    )
+
+    dockerfile_path = tmp_dir / "Dockerfile"
+    dockerfile_path.write_text(docker_content)
+
+
+class DefaultImageBuilder(ImageSpecBuilder):
+    """Image builder using Docker and buildkit."""
+
+    _SUPPORTED_IMAGE_SPEC_PARAMETERS: ClassVar[set] = {
+        "name",
+        "python_version",
+        "builder",
+        "source_root",
+        "env",
+        "registry",
+        "packages",
+        "conda_packages",
+        "conda_channels",
+        "requirements",
+        "apt_packages",
+        "platform",
+        "cuda",
+        "cudnn",
+        "base_image",
+        "pip_index",
+        # "registry_config",
+        "commands",
+    }
+
+    def build_image(self, image_spec: ImageSpec) -> str:
+        return self._build_image(image_spec)
+
+    def _build_image(self, image_spec: ImageSpec, *, push: bool = True) -> str:
```

Comment:
I've been using this private method to build images locally and not push. It's been helpful to debug the image without having to wait for a push.